runners/flink/src/main/java/org/apache/beam/runners/flink/translation/wrappers/streaming/ExecutableStageDoFnOperator.java

```diff
@@ -544,30 +601,49 @@ public void processWatermark(Watermark mark) throws Exception {
     // every watermark. So we have implemented 2) below.
     //
     if (sdkHarnessRunner.isBundleInProgress()) {
-      if (mark.getTimestamp() >= BoundedWindow.TIMESTAMP_MAX_VALUE.getMillis()) {
-        invokeFinishBundle();
-        setPushedBackWatermark(Long.MAX_VALUE);
+      if (minEventTimeTimerTimestampInLastBundle < Long.MAX_VALUE) {
+        // We can safely advance the watermark to before the last bundle's minimum event timer
+        // but not past the potential output watermark which includes holds to the input watermark.
+        return Math.min(minEventTimeTimerTimestampInLastBundle - 1, potentialOutputWatermark);
       } else {
-        // It is not safe to advance the output watermark yet, so add a hold on the current
-        // output watermark.
-        backupWatermarkHold = Math.max(backupWatermarkHold, getPushbackWatermarkHold());
-        setPushedBackWatermark(Math.min(currentOutputWatermark, backupWatermarkHold));
-        super.setBundleFinishedCallback(
-            () -> {
-              try {
-                LOG.debug("processing pushed back watermark: {}", mark);
-                // at this point the bundle is finished, allow the watermark to pass
-                // we are restoring the previous hold in case it was already set for side inputs
-                setPushedBackWatermark(backupWatermarkHold);
-                super.processWatermark(mark);
-              } catch (Exception e) {
-                throw new RuntimeException(
-                    "Failed to process pushed back watermark after finished bundle.", e);
-              }
-            });
+        // We don't have any information yet, use the current output watermark for now.
+        return currentOutputWatermark;
+      }
+    } else {
+      // No bundle was started when we advanced the input watermark.
+      // Thus, we can safely set a new output watermark.
+      return potentialOutputWatermark;
+    }
+  }
+
+  private void preBundleStartCallback() {
+    inputWatermarkBeforeBundleStart = getEffectiveInputWatermark();
+  }
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  private void finishBundleCallback() {
+    minEventTimeTimerTimestampInLastBundle = minEventTimeTimerTimestampInCurrentBundle;
+    minEventTimeTimerTimestampInCurrentBundle = Long.MAX_VALUE;
+    try {
+      if (!closed
+          && minEventTimeTimerTimestampInLastBundle < Long.MAX_VALUE
+          && minEventTimeTimerTimestampInLastBundle <= getEffectiveInputWatermark()) {
+        ProcessingTimeService processingTimeService = getProcessingTimeService();
+        // We are scheduling a timer for advancing the watermark. Otherwise we
+        // could potentially loop forever here when a timer keeps scheduling a timer
+        // for the same timestamp. This in itself would not be an issue. However,
```

Comment:
The comment is a bit misleading. How about: "We are scheduling a timer for advancing the watermark, to not delay finishing the bundle and temporarily release the checkpoint lock. Otherwise, we could potentially loop when a timer keeps scheduling a timer for the same timestamp."

Reply:
Sure. Changed.